### ISSUE TITLE
Cypress will not stop test run if errors in beforeEach/afterEach

### ISSUE THREAD

**Issue by tomi-roam** [cypress-io/cypress#674]
- Operating System: MacOS and Docker image (`cypress/base:8` and `node:8.5`)
- Cypress Version: 0.20.1
- Browser Version: Electron

### Is this a Feature or Bug?
Bug

### Current behavior:
If an error is raised in both the `beforeEach` and `afterEach` hooks for tests, Cypress will not stop the test run and seems to continue recording video indefinitely (video file keeps increasing).

### Desired behavior:
Tests should fail with a descriptive error message.

### How to reproduce:
Run the test below.

`node_modules/.bin/cypress run` will "hang" and continue running until killed
`node_modules/.bin/cypress run --headed` will also "hang", but you can see the errors in the log and see the timer continuing to run.

#### Test code:
```js
describe("cypress", function() {
  beforeEach(function() {
    // cy.visit("/");
    throw new Error();
  });

  afterEach(function() {
    // Commenting this out will allow the test to finish and stop recording
    // cy.visit("/");
    throw new Error();
  });

  it("doesn't hang when both beforeEach and afterEach fail", function() {
  });
});
```

### Additional Info (images, stack traces, etc)
- The original test makes a `cy.visit` call and the tests would hang if the server didn't respond with a 200.
- Canceling the `--headed` run results in output similar to the following:
```
  cypress
    1) "before each" hook for "doesn't hang when both beforeEach and afterEach fail"

  0 passing (15s)
  1 failing

  1) cypress "before each" hook for "doesn't hang when both beforeEach and afterEach fail":
     E

Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 'cypress'


  Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 'cypress'r

  Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 'cypress'o

  Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 'cypress'r

  Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 'cypress'
```

Looking at the "random" character after 'cypress', it starts to spell out:
`Error at Context. <anonymous> (http://localhost:64422/__cypress/tests` ...

**Comment by brian-mann** [comment]
Can confirm. The problem is when both the `beforeEach` and `afterEach` fail it causes Cypress not to exit.

**Comment by brian-mann** [comment]
I am not sure how this slipped through the cracks but it should have been fixed awhile ago. I'll put it on a milestone so that we actually fix it...

**Comment by bahmutov** [comment]
👍 we got to do code reviews to maybe spot problematic code paths

**Comment by brian-mann** [comment]
I'll report back on when I find the root cause. It could also be a problem with mocha itself. I haven't tried to reproduce again (since September) and it's possible maybe this is already fixed...

**Comment by brian-mann** [comment]
What's even weird about this bug is that it only happens when running `cypress run` and it only happens if both of the errors have no message body. Passing `throw new Error('foo')` works.

**Comment by brian-mann** [comment]
Figured out the error part - this is actually a problem with the reporter.

When there are no commands in a test AND no properties change the reporter will not update, and Cypress will hang forever.

Because the error messages were the same (an empty string) `""`, and the test was already `isOpen` due to not having any commands - it would never update.

**Comment by brian-mann** [comment]
The fundamental problem about this isn't about errors - it's about state. When a test has its state updated and no properties actually change (due to a perfect storm of problems) - the reporter will never update the UI, and Cypress will never take a screenshot and continue on. That's the area it's failing at.

**Comment by brian-mann** [comment]
Released in [`2.0.2`](https://docs.cypress.io/guides/references/changelog.html#2-0-2).
